python/revng/model/metaaddress.py

```diff
@@ -0,0 +1,162 @@
+import re
+from dataclasses import dataclass, field
+from enum import Enum
+from typing import Type
+
+import yaml
+
+
+class MetaAddressType(Enum):
+    Invalid = "Invalid"
+    Generic32 = "Generic32"
+    Generic64 = "Generic64"
+    Code_x86 = "Code_x86"
+    Code_x86_64 = "Code_x86_64"
+    Code_mips = "Code_mips"
+    Code_mipsel = "Code_mipsel"
+    Code_arm = "Code_arm"
+    Code_arm_thumb = "Code_arm_thumb"
+    Code_aarch64 = "Code_aarch64"
+    Code_systemz = "Code_systemz"
+
+    @classmethod
+    def yaml_representer(cls, dumper: yaml.dumper.Dumper, instance: Enum):
+        return dumper.represent_str(instance.name)
+
+
+@dataclass
+class MetaAddress:
+    Address: int
+    Type: MetaAddressType
+    Epoch: int = field(default=0)
+    AddressSpace: int = field(default=0)
+
+    def __init__(self, *args, **kwargs):
```

Comment:
This is a quite obscure constructor and lacks docs.
Can't we have a simple constructor and maybe one or more `static` methods with explicit names and simple parameter list? For instance `MetaAddress.fromString` or the like.

Reply:
Yeah for sure the constructor needs documentation. I took the decision to implement a polymorphic constructor because it's what I would like to have as a python user, given that MetaAddress is pervasive I would like to have the convenience of being able to do all of the following without hassle
```python
MetaAddress("0x0:Invalid") # construction from serialized form
MetaAddress(0x0, "Invalid")
MetaAddress(0x0, MetaAddressType.Invalid)
MetaAddress(0x0, Type=MetaAddressType.Invalid)
MetaAddress(0x0, MetaAddressType.Invalid, 123, 456) # 123 is the epoch, 456 the address space
MetaAddress(0x0, MetaAddressType.Invalid, Epoch=123, AddressSpace=456)
MetaAddress(Address=0x0, Type=MetaAddressType.Invalid, Epoch=123, AddressSpace=456)
kwargs = {"Address": 0, Type="Invalid"} # this could come from a JSON
MetaAddress(**kwargs)
```
Having factory methods is pretty inconvenient IMO as long as the constructor is documented and predictable, but I recognize it's an opinionated opinion :)

There's another reason why a polymorphic constructor is "required", namely to not special case it in the YAML deserialization process, as the YAML constructor function invokes the regular constructor for all types that are not part of the schema.

Reply:
This is quite complicated to implement and document. Having `MetaAddress.fromString` is obvious and self-documenting.
I'd have a constructor that accepts `kwargs` and a `MetaAddress.fromString` `@classmethod` and be done with it.
KISS.